drivers/wifi/esp/esp_offload.c

```diff
@@ -365,30 +359,10 @@ static int esp_sendto(struct net_pkt *pkt,
 		}
 	}
 
-	sock->tx_pkt = pkt;
-	sock->send_cb = cb;
-	sock->send_user_data = user_data;
-
-	if (timeout == 0) {
-		k_work_submit_to_queue(&dev->workq, &sock->send_work);
-		return 0;
-	}
-
-	ret = _sock_send(dev, sock);
-	if (ret == 0) {
-		net_pkt_unref(sock->tx_pkt);
-	} else {
-		LOG_ERR("Failed to send data: link %d, ret %d", sock->link_id,
-			ret);
-	}
-
-	sock->tx_pkt = NULL;
-
-	if (cb) {
-		cb(context, ret, user_data);
-	}
+	k_work_init(&pkt->work, esp_send_work);
+	k_work_submit_to_queue(&dev->workq, &pkt->work);
```

Comment:
Doesn't this remove the possibility of doing a blocking `send()`?

Reply:
Well, `send()` will still work, but it will not block on the function. Instead, data will be scheduled for sending, which is similar to what happens within native network stack. There is no way to block until packet will go through the wire or reach the target. So all that we change is whether packet will be scheduled at ESP chip side or already at Zephyr side (and then at the ESP side as well).

Reply:
Right, but any errors from `_sock_send()` won't be returned? I'm not saying this is incorrect but it will affect the behavior somewhat. Before it was that when a send (with timeout !=0) returns it has been delivered to the ESP and now it has been put on the internal queue _to be_ delivered.

Reply:
So yes, behavior changes. Errors will not be returned anymore, similarly as timeout==0 was working so far. This driver is one of the few to execute different code path depending on blocking or non-blocking sockets. I think that this originate from a bugfix one of network offloaded drivers and was copy pasted to the others. In general I don't think this behavior was correct, so this PR includes proposal for changing it.

As you say, behavior will be different, but not wrong. The gain here is that we remove one code path.